pkg/proxy/proxy.go

```diff
@@ -149,6 +168,78 @@ func (p *Proxy) StartProxy() *http.Server {
 	return srv
 }
 
+// unsecured returns true if the request does not require authentication
+func unsecured(ctx echo.Context) bool {
+	uri := ctx.Request().URL.RequestURI()
+	return uri == proxyHealthEndpoint || uri == "/.well-known/oauth-authorization-server" || strings.HasPrefix(uri, authEndpoint)
+}
+
+// auth handles requests to SSO. Used by web login.
+func (p *Proxy) auth(ctx echo.Context) error {
+	req := ctx.Request()
+	targetURL, err := url.Parse(configuration.GetRegistrationServiceConfig().Auth().SSOBaseURL())
+	if err != nil {
+		return err
+	}
+	targetURL.Path = req.URL.Path
+	targetURL.RawQuery = req.URL.RawQuery
+
+	return p.handleSSORequest(targetURL)(ctx)
+}
+
+// oauthConfiguration handles requests to oauth configuration and proxies them to the corresponding SSO endpoint. Used by web login.
+func (p *Proxy) oauthConfiguration(ctx echo.Context) error {
+	targetURL, err := url.Parse(ssoWellKnownTarget())
+	if err != nil {
+		return err
+	}
+	return p.handleSSORequest(targetURL)(ctx)
+}
+
+// openidAuth handles requests to the openID Connect authentication endpoint. Used by web login.
+func (p *Proxy) openidAuth(ctx echo.Context) error {
+	targetURL, err := url.Parse(authorizationEndpointTarget())
+	if err != nil {
+		return err
+	}
+	targetURL.Path = ctx.Request().URL.Path
+	targetURL.RawQuery = ctx.Request().URL.RawQuery
+
+	return p.redirectTo(ctx, targetURL.String())
+}
+
+func (p *Proxy) redirectTo(ctx echo.Context, to string) error {
+	log.InfoEchof(ctx, "redirecting %s to %s", ctx.Request().URL.String(), to)
+	http.Redirect(ctx.Response().Writer, ctx.Request(), to, http.StatusSeeOther)
+	return nil
+}
+
+// handleSSORequest handles requests to the cluster authentication server and proxy them to SSO instead. Used by web login.
+func (p *Proxy) handleSSORequest(targetURL *url.URL) echo.HandlerFunc {
+	return func(ctx echo.Context) error {
+		req := ctx.Request()
+		director := func(req *http.Request) {
+			origin := req.URL.String()
+			req.URL.Scheme = targetURL.Scheme
+			req.URL.Host = targetURL.Host
```

Comment:
curious if there is a difference between req.URL.Host and req.Host? both have the same value being assigned i.e targetURL.Host

Reply:
Surprisingly there seems to be some difference. Setting req.URL.Host is enough for some services (Kube API Server, our own Dev Keycloak). But Akamai which is used by sso.redhat.com also requires req.Host to be set to the same value as the req.URL.Host.